lib/multi-line/smart-multi-line.c

```diff
@@ -0,0 +1,453 @@
+                 gboolean *segment_is_part_of_trace,
+                 gboolean *segment_starts_a_new_trace,
+                 gboolean *segment_ends_trace)
+{
+  *segment_is_part_of_trace = FALSE;
+  *segment_starts_a_new_trace = FALSE;
+  *segment_ends_trace = FALSE;
+
+  gboolean last_segment_ended_the_trace = self->current_state == SMLS_START_STATE;
+
+  gboolean segment_is_trace = _fsm_transition(self, segment, segment_len);
+
+  msg_trace_printf("smart-multi-line: [STEP1] >>%.*s<<, result=%d, state=%d", (int) segment_len, segment,
+                   segment_is_trace, self->current_state);
+  *segment_is_part_of_trace = segment_is_trace;
+
+  if (!(*segment_is_part_of_trace))
+    {
+      /* try again from the start state, the current segment is may be part of a new trace */
+      segment_is_trace = _fsm_transition(self, segment, segment_len);
+
+      msg_trace_printf("smart-multi-line: [STEP2]: >>%.*s<<, result=%d, state=%d", (int) segment_len, segment,
+                       segment_is_trace, self->current_state);
+
+      *segment_is_part_of_trace = segment_is_trace;
+      if (*segment_is_part_of_trace)
+        *segment_starts_a_new_trace = TRUE;
+    }
+  else
+    {
+      if (last_segment_ended_the_trace)
+        *segment_starts_a_new_trace = TRUE;
+      *segment_ends_trace = (self->current_state == SMLS_START_STATE);
+    }
+}
+
+static gint
+_accumulate_line_unlocked(MultiLineLogic *s,
+                          const guchar *msg, gsize msg_len,
+                          const guchar *segment, gsize segment_len)
+{
+  SmartMultiLine *self = (SmartMultiLine *) s;
+
+  if (self->last_segment_rewound)
+    {
+      /* we rewound the last segment, processing it again.  The previous
+       * message was extracted we are starting with a fresh start, but we
+       * don't run the current segment through the FSM, we simply reuse the
+       * result the last time we've run it.  */
+
+      g_assert(msg_len == 0);
+
+      self->last_segment_rewound = FALSE;
+      if (self->rewound_segment_is_trace)
+        {
+          self->consumed_message_is_trace = TRUE;
+          return MLL_CONSUME_SEGMENT | MLL_WAITING;
+        }
+      else
+        {
+          self->consumed_message_is_trace = FALSE;
+          return MLL_CONSUME_SEGMENT | MLL_EXTRACTED;
+        }
+    }
+  else
+    {
+      gboolean segment_is_part_of_trace, segment_starts_a_new_trace, segment_ends_trace;
+
+      _process_segment(self, (const gchar *) segment, segment_len,
+                       &segment_is_part_of_trace,
+                       &segment_starts_a_new_trace,
+                       &segment_ends_trace);
+
+      if (msg_len == 0)
+        {
+          /* just starting up, first segment is to be consumed */
+
+          if (!segment_is_part_of_trace)
+            {
+              /* not recognized as part of a trace, just return it */
+              return MLL_CONSUME_SEGMENT | MLL_EXTRACTED;
+            }
```

Comment:
since we return MLL_EXTRACTED here, no msg remains partially consumed. But I agree, it might be clearer, so I added an explicit statement to set consumed_message_is_trace to FALSE.

note that this is not going to be used, at least with the current code.

Also note, that consumed_message_is_trace refers to the partially consumed message while segment_is_part_of_trace refers only to the current segment and shows whether it is considered to be a continuation.